assets/src/models/createDetourMachine.ts

```diff
@@ -1,23 +1,114 @@
+      | { type: "detour.edit.undo" }
+      | { type: "detour.edit.done" }
+      | { type: "detour.share.copy-detour"; detourText: string }
+  },
+  actions: {
+    "detour.add-waypoint": assign({
+      waypoints: ({ context }, params: { location: ShapePoint }) => [
+        ...context.waypoints,
+        params.location,
+      ],
+    }),
+    "detour.clear": assign({
+      waypoints: [],
+    }),
+    "detour.undo": assign({
+      waypoints: ({ context }) => context.waypoints.slice(0, -1),
+    }),
+    "detour.share.copy-detour": (_, { detourText }: { detourText: string }) => {
+      window.navigator.clipboard?.writeText(detourText)
+    },
   },
 }).createMachine({
   id: "Detours Machine",
+  context: () => ({
+    waypoints: [],
+  }),
 
   initial: "Detour Drawing",
   states: {
     "Detour Drawing": {
       initial: "Editing",
       states: {
+        // ? Can this be modeled with a sub state machine as an actor?..
         Editing: {
-          initial: "Idle",
+          initial: "Pick Start Point",
+          on: {
+            "detour.edit.clear-detour": {
+              target: "Editing",
+              actions: "detour.clear",
+            },
+          },
           states: {
-            Idle: {
+            "Pick Start Point": {
+              on: {
+                "detour.edit.place-connection-point": {
+                  target: "Place Waypoint",
+                  actions: [
+                    {
+                      type: "detour.add-waypoint",
+                      params: ({ event }) => ({
+                        location: event.location,
+                      }),
+                    },
+                  ],
+                },
+              },
+            },
+            "Place Waypoint": {
               on: {
-                "detour.share": {
+                "detour.edit.place-waypoint": {
+                  actions: [
+                    {
+                      type: "detour.add-waypoint",
+                      params: ({ event }) => ({
+                        location: event.location,
+                      }),
+                    },
+                  ],
+                },
+                "detour.edit.place-connection-point": {
+                  target: "Finished Drawing",
+                  actions: [
+                    {
+                      type: "detour.add-waypoint",
+                      params: ({ event: { location } }) => ({
+                        location,
+                      }),
+                    },
+                  ],
+                },
+                "detour.edit.undo": [
+                  {
+                    guard: ({ context }) => context.waypoints.length === 1,
+                    actions: "detour.undo",
+                    target: "Pick Start Point",
+                  },
+                  { actions: "detour.undo", target: "Place Waypoint" },
+                ],
+              },
+            },
+            "Finished Drawing": {
```

Comment:
Hmm, I don't think I understand what you mean by "Connect endpoint" here? The only actions left to do at this state is "finishing" or "undo"-ing, so I don't think it applies because there's nothing to "Connect"? Maybe I'm misunderstanding?

Reply:
Oooooo... I'm not sure what I was meaning here anymore! Agree that the current naming makes sense as-is. (I think I wrote the above when I was trying to understand the differences between some states and events?)